### PR TITLE
Implement new protocol (#25)

### LINKED ISSUE #25: Implement new protocol
A new protocol for WebSocket messages was created to handle concurrency issues with `Cursor Update` messages. It is outlined [here](https://github.com/schramm-famm/docs/blob/master/design/websocket_protocol.md#message-protocol) in the `Server` section. `Sync` type messages also must be added.

### PR DESCRIPTION
**Pull request by Thao-Tran**
Resolves #25 
* Add Sync message handling
* Add checkpoint field to conversations to track user cursors at each
version

### CODE REVIEW

File: websockets/conversation.go [hunk 1]
```diff
@@ -154,11 +178,28 @@ func (c *Conversation) handleEditUpdate(msg protocol.Message, sender *Client) er
...
 		if client != sender {
 			client.caret = client.caret.ShiftCaret(sender.caret, *update.Delta)
+			newCheckpoint.syncsLeft[client.userID] = true
+		}
+		newCheckpoint.activeUsers[client.userID] = client.caret
+	}
+
+	c.checkpoint[*update.Version] = newCheckpoint
+
+	if len(newCheckpoint.syncsLeft) == 0 {
```
Comment (igorveselinovic) [inline websockets/conversation.go:189]: This only gets triggered in the scenario where only one user has the conversation open right?

File: websockets/conversation.go [hunk 2]
```diff
@@ -146,6 +163,13 @@ func (c *Conversation) handleEditUpdate(msg protocol.Message, sender *Client) er
...
 	}
 
+	newCheckpoint := &Checkpoint{
+		activeUsers: make(map[int64]protocol.Caret),
+		senderCaret: sender.caret,
+		delta:       *update.Delta,
+		syncsLeft:   make(map[int64]bool),
+	}
+
 	// Update the sender's caret
```
Comment (igorveselinovic) [inline websockets/conversation.go]: I mentioned this in the PR for the WebSocket protocol doc: I think that applying the delta to the sender's cursor needs to be done *after* the shifting of all the other users' cursors. I did the steps in the wrong order when I wrote it.

**Review by igorveselinovic — CHANGES_REQUESTED** [review]
I think we might also need to update the `syncsLeft` map in every `Checkpoint` whenever a user leaves a conversation (i.e., whenever `unregisterClient` is called). 

It would be possible for a client to receive a broadcasted update, then close the tab which would result in them never sending back a `Sync` message.

File: websockets/conversation.go (hunk 1 above)
Comment (Thao-Tran) [reply websockets/conversation.go:189]: Right

**Comment by Thao-Tran** [comment]
> I think we might also need to update the `syncsLeft` map in every `Checkpoint` whenever a user leaves a conversation (i.e., whenever `unregisterClient` is called).
> 
> It would be possible for a client to receive a broadcasted update, then close the tab which would result in them never sending back a `Sync` message.

Good call. Yeah I'll add that then.

**Review by igorveselinovic — APPROVED** [review]

**Pull request merged**